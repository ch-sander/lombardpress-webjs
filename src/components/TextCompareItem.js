import React from 'react';
import Diff from 'diff-match-patch'
import Axios from 'axios'

import { FaEyeSlash, FaEye, FaStar } from 'react-icons/fa';

class TextCompareItem extends React.Component {
  constructor(props){
    super(props)
    this.handleToggleShow = this.handleToggleShow.bind(this)
    this.mounted = ""
    this.state = {
      compareText: "",
      rawText: "",
      show: true
    }
  }
  handleToggleShow(){
    this.setState((prevState) => {
      return{
        show: !prevState.show
      }
    })
  }
  createCompare(base, transcription){

    Axios.get("http://exist.scta.info/exist/apps/scta-app/csv-pct.xq?resourceid=" + transcription).
          then((text) => {

            const dmp = new Diff.diff_match_patch();

            const diff = dmp.diff_main(base, text.data);
            // Result: [(-1, "Hell"), (1, "G"), (0, "o"), (1, "odbye"), (0, " World.")]
            dmp.diff_cleanupSemantic(diff);
            const ds = dmp.diff_prettyHtml(diff);
            if (this.mounted === true && base){
              this.setState({compareText: ds, rawText: text.data})
            }

          })
        }

  componentDidMount(){
    this.mounted = true;
    this.setState({rawText: "", compareText: ""})
    this.createCompare(this.props.base, this.props.compareTranscription)
  }
  componentWillReceiveProps(newProps){
    // conditional try to restrict new async calls to only when props.info changes
    if (newProps.base != this.props.base || newProps.compareTranscription != this.props.compareTranscription){
      this.setState({rawText: "", compareText: ""})
      this.createCompare(newProps.base, newProps.compareTranscription)
    }
  }
  componentWillUnmount(){
      this.mounted = false;
    }


  render(){
    const displayComparison = () => {
      if (this.props.base && this.state.compareText){
        return (
          <div>
            <span onClick={() => this.handleToggleShow()}>{this.state.show ? <FaEyeSlash/> : <FaEye/>}</span>
            <span onClick={() => this.props.handleChangeBase(this.state.rawText)}><FaStar/></span>
            <div className={this.state.show ? "unhidden" : "hidden"}>
              <div ref="text" dangerouslySetInnerHTML={{ __html: this.state.compareText}}></div>
            </div>
          </div>
        )
      }
      else{
        return <p>Loading</p>
      }
    }

    return (
      displayComparison()
    );
  }
}

export default TextCompareItem;
