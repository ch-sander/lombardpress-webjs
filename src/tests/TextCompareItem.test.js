import React from 'react';
import ReactDOM from 'react-dom';
import {shallow} from 'enzyme'

import TextCompareItem from '../components/TextCompareItem';


it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<TextCompareItem />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('renders without crashing when props are present', () => {
  const div = document.createElement('div');
  ReactDOM.render(<TextCompareItem 
    base="circa prologum sententiarum in quo magister dicit quod intentionis suae est munire davidicam turrim vel potius munitam ostendere clypeis etc quaero istam quaestionem utrum in causa iudiciali fidei contra traditionem pure humanitus adinventam iudex idoneus ferret pro fide sententiam"
    compareTranscription="http://scta.info/resource/l1-cpspfs/reims/transcription"
    show={true}
    showCompare={true}
    />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('should show no children because prop show is set to false', () => {
  const wrapper = shallow(<TextCompareItem 
    base="circa prologum sententiarum in quo magister dicit quod intentionis suae est munire davidicam turrim vel potius munitam ostendere clypeis etc quaero istam quaestionem utrum in causa iudiciali fidei contra traditionem pure humanitus adinventam iudex idoneus ferret pro fide sententiam"
    compareTranscription="http://scta.info/resource/l1-cpspfs/reims/transcription"
    show={false}
    showCompare={true}
    />)
  expect(wrapper.find('div').length).toBe(0)
});

it('should immediately render with Spinner as async call is started', () => {
  const wrapper = shallow(<TextCompareItem 
    base="circa prologum sententiarum in quo magister dicit quod intentionis suae est munire davidicam turrim vel potius munitam ostendere clypeis etc quaero istam quaestionem utrum in causa iudiciali fidei contra traditionem pure humanitus adinventam iudex idoneus ferret pro fide sententiam"
    compareTranscription="http://scta.info/resource/l1-cpspfs/reims/transcription"
    show={true}
    showCompare={true}
    />)
  expect(wrapper.find('Spinner').length).toBe(1)
});
///TODO need test for after async call is finished; and spinner should be removed and replaced with div
