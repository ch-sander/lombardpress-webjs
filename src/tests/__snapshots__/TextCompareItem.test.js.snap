// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`set states and displays compare text when showCompare is changed to true 1`] = `
<div
  style={
    Object {
      "borderLeft": "5px solid rgba(NaN, 0, 255, 1)",
      "paddingLeft": "5px",
    }
  }
>
  <span>
    <Link
      to="/text?resourceid=http://scta.info/resource/l1-cpspfs/reims/transcription"
    >
      Paragraph l1-cpspfs/reims
    </Link>
     
  </span>
  <span
    className="lbp-span-link"
    onClick={[Function]}
    title="show/hide"
  >
    <FaEyeSlash />
  </span>
  <span
    className="lbp-span-link"
    onClick={[Function]}
    title="toggle comparison off"
  >
    <FaToggleOn />
  </span>
  <span
    className="lbp-span-link"
    onClick={[Function]}
    title="set as base"
  >
    <FaStar />
  </span>
  <span
    style={
      Object {
        "fontSize": "14px",
        "paddingLeft": "4px",
      }
    }
    title="levenshtein distance"
  />
  <div
    className="unhidden"
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "circa prologum sententiarum in quo magister dicit quod intentionis suae est munire davidicam turrim vel potius munitam ostendere clypeis etc quaero istam quaestionem utrum in causa iudiciali fidei contra traditionem pure humanitus adinventam iudex idoneus ferret pro fide sententiam",
        }
      }
    />
    <div>
      <span
        onClick={[Function]}
      >
        <FaRegImage />
      </span>
    </div>
  </div>
</div>
`;

exports[`set states and displays raw text when showCompare is changed to False 1`] = `
<div>
  <span>
    <Link
      to="/text?resourceid=http://scta.info/resource/l1-cpspfs/reims/transcription"
    >
      Paragraph l1-cpspfs/reims
    </Link>
  </span>
   |
  <span
    className="lbp-span-link"
    onClick={[Function]}
    title="show/hide"
  >
    <FaEyeSlash />
  </span>
  <span
    className="lbp-span-link"
    onClick={[Function]}
    title="toggle comparison on"
  >
    <FaToggleOff />
  </span>
  <span
    className="lbp-span-link"
    onClick={[Function]}
    title="set as base"
  >
    <FaStar />
  </span>
  <div
    className="unhidden"
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "TEST Circa prologum sententiarum in quo magister dicit quod intentionis suae est munire davidicam turrim vel potius munitam ostendere clypeis etc quaero istam quaestionem utrum in causa iudiciali fidei contra traditionem pure humanitus adinventam iudex idoneus ferret pro fide sententiam",
        }
      }
    />
    <div>
      <span
        onClick={[Function]}
      >
        <FaRegImage />
      </span>
    </div>
  </div>
</div>
`;
